gcp/api/server.py

```diff
@@ -1080,68 +1181,81 @@ def query_by_version(context: QueryContext,
-                                                    purl, version)
+      # If the previous query has fully finished (or skipped),
+      # try generic version
+      new_bugs = yield _query_by_generic_version(context, query, package_name,
+                                                 ecosystem, purl, version)
       for bug in new_bugs:
         if bug not in bugs:
           bugs.append(bug)
+
     elif project == 'oss-vdb-test' and supports_ordering:
       # Query for non-enumerated ecosystems.
 
-      bugs, next_page_token = yield _query_by_comparing_versions(
-          context, query, ecosystem, version)
+      bugs = yield _query_by_comparing_versions(context, query, ecosystem,
+                                                version)
     else:
-      bugs, next_page_token = yield _query_by_generic_version(
-          context, query, package_name, ecosystem, purl, version)
+      bugs = yield _query_by_generic_version(context, query, package_name,
+                                             ecosystem, purl, version)
 
   else:
     logging.warning("Package query without ecosystem specified")
     # Unspecified ecosystem. Try semver first.
-
-    # TODO: Remove after testing how many consumers are
-    # querying the API this way.
-    context.page_token = None
-    new_bugs, _ = yield _query_by_semver(context, query, package_name,
-                                         ecosystem, purl, version)
+    new_bugs = yield _query_by_semver(context, query, package_name, ecosystem,
+                                      purl, version)
     bugs.extend(new_bugs)
 
-    new_bugs, _ = yield _query_by_generic_version(context, query, package_name,
-                                                  ecosystem, purl, version)
+    new_bugs = yield _query_by_generic_version(context, query, package_name,
+                                               ecosystem, purl, version)
     for bug in new_bugs:
       if bug not in bugs:
         bugs.append(bug)
 
-    # Trying both is too difficult/ugly with paging
     # Our documentation states that this is an invalid query
     # context.service_context.abort(grpc.StatusCode.INVALID_ARGUMENT,
     #                               'Ecosystem not specified')
 
-  return [to_response(bug) for bug in bugs], next_page_token
+  return [to_response(bug) for bug in bugs]
 
 
 @ndb.tasklet
 def _query_by_comparing_versions(context: QueryContext, query: ndb.Query,
-                                 ecosystem: str,
-                                 version: str) -> tuple[list, ndb.Cursor]:
-  """Query by package."""
+                                 ecosystem: str, version: str) -> list:
+  """
+  Query by comparing versions.
+
+  Args:
+    context: QueryContext for the current query.
+    query: A partially completed ndb.Query object which only needs 
+      version filters to be added before query is performed.
+    ecosystem: Optional ecosystem of the package to query.
```

Comment:
For comparing_versions, `ecosystem` is required. It needs `ecosytem.sort_key()` to compare versions.

Reply:
Updated